From f225dda8654ffd9fd2b9d425848007cbdfa17b32 Mon Sep 17 00:00:00 2001
From: Bharat Bhushan <Bharat.Bhushan@nxp.com>
Date: Mon, 6 Mar 2017 15:22:00 +0530
Subject: [PATCH 528/706] vfio/fsl-mc: non-dprc device support added

Non-DPRC devices shares it's parent container MC portal
to communicate to MC hardware.

Signed-off-by: Bharat Bhushan <Bharat.Bhushan@nxp.com>
(cherry picked from commit 847f513bf8e119286d88dc23c231ad158de76943)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/vfio/fsl-mc/vfio_fsl_mc.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/drivers/vfio/fsl-mc/vfio_fsl_mc.c b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
index 8f3a62541e72..eb4024922a9c 100644
--- a/drivers/vfio/fsl-mc/vfio_fsl_mc.c
+++ b/drivers/vfio/fsl-mc/vfio_fsl_mc.c
@@ -229,9 +229,16 @@ static int vfio_fsl_mc_probe(struct fsl_mc_device *mc_dev)
 			goto free_vfio_device;
 		}
 	} else {
-		/* Handling for Non-DPRC device to be added */
-		ret = -EINVAL;
-		goto free_vfio_device;
+		struct fsl_mc_device *mc_bus_dev;
+
+		/* Non-dprc devices share mc_io from the parent dprc */
+		mc_bus_dev = to_fsl_mc_device(mc_dev->dev.parent);
+		if (mc_bus_dev == NULL) {
+			vfio_del_group_dev(dev);
+			goto free_vfio_device;
+		}
+
+		mc_dev->mc_io = mc_bus_dev->mc_io;
 	}
 	return 0;
 
@@ -253,6 +260,8 @@ static int vfio_fsl_mc_remove(struct fsl_mc_device *mc_dev)
 	if (strcmp(mc_dev->obj_desc.type, "dprc") == 0)
 		vfio_fsl_mc_cleanup_dprc(vdev);
 
+	mc_dev->mc_io = NULL;
+
 	vfio_iommu_group_put(mc_dev->dev.iommu_group, dev);
 	kfree(vdev);
 
-- 
2.17.1

